DataDog/libdatadog
# feat(common): add tag! macro

# What does this PR do?

This does a few things:

 1. Creates a `tag!` macro which does compile-time checking for some invariants.
 2. Changes a `Tag` to use `Cow<'static, str>` instead of a `String`.
 3. Makes `Tag::from_value` private. Use `tag!` if the key and value are known at compile-time. If not, use `Tag::new`.
 4. `Tag::new` returns an `anyhow::Result<Tag>` now instead of a `Result<Tag, Cow<_>>`.

# Motivation

As libdatadog usage has grown, we have more Rust users of the lib and therefore more static Rust strs we can trust at compile-time and avoid runtime allocations.

# Additional Notes

This was previously stacked on another PR, but not anymore.

# How to test the change?

Everything should run as normal. It's recommended to use `tag!` over `Tag::new` if the key and value are both static strs.
